src/source/custom_source.js

```diff
@@ -0,0 +1,319 @@
+ *         const imageBitmap = await window.createImageBitmap(blob);
+ *
+ *         return imageBitmap;
+ *     }
+ * }
+ *
+ * map.on('load', () => {
+ *     map.addSource('custom-source', new CustomSource());
+ *     map.addLayer({
+ *         id: 'layer',
+ *         type: 'raster',
+ *         source: 'custom-source'
+ *     });
+ * });
+ */
+
+/**
+ * Optional method called when the source has been added to the Map with {@link Map#addSource}.
+ * This gives the source a chance to initialize resources and register event listeners.
+ *
+ * @function
+ * @memberof CustomSourceInterface
+ * @instance
+ * @name onAdd
+ * @param {Map} map The Map this custom source was just added to.
+ */
+
+/**
+ * Optional method called when the source has been removed from the Map with {@link Map#removeSource}.
+ * This gives the source a chance to clean up resources and event listeners.
+ *
+ * @function
+ * @memberof CustomSourceInterface
+ * @instance
+ * @name onRemove
+ * @param {Map} map The Map this custom source was added to.
+ */
+
+/**
+ * Optional method called after the tile is unloaded from the map viewport. This
+ * gives the source a chance to clean up resources and event listeners.
+ *
+ * @function
+ * @memberof CustomSourceInterface
+ * @instance
+ * @name unloadTile
+ * @param {{ z: number, x: number, y: number }} tile Tile name to unload in the XYZ scheme format.
+ */
+
+/**
+ * Optional method called during a render frame to allow a source to prepare and modify a tile texture if needed.
+ *
+ * @function
+ * @memberof CustomSourceInterface
+ * @instance
+ * @name prepareTile
+ * @param {{ z: number, x: number, y: number }} tile Tile name to prepare in the XYZ scheme format.
+ * @returns {TextureImage} The tile image data as an `HTMLImageElement`, `ImageData`, `ImageBitmap` or object with `width`, `height`, and `data`.
+ */
+
+/**
+ * Called when the map starts loading tile for the current animation frame.
+ *
+ * @function
+ * @memberof CustomSourceInterface
+ * @instance
+ * @name loadTile
+ * @param {{ z: number, x: number, y: number }} tile Tile name to load in the XYZ scheme format.
+ * @param {Object} options Options.
+ * @param {AbortSignal} options.signal A signal object that allows the map to cancel tile loading request.
+ * @returns {Promise<TextureImage>} The tile image data as an `HTMLImageElement`, `ImageData`, `ImageBitmap` or object with `width`, `height`, and `data`.
+ */
+export type CustomSourceInterface<T> = {
+    id: string;
+    type: 'custom',
+    dataType: ?DataType,
+    minzoom: number,
+    maxzoom: number,
+    scheme: string;
+    tileSize: number,
+    attribution: ?string,
+    loadTile: (tileID: { z: number, x: number, y: number }, options: { signal: AbortSignal }) => Promise<T>,
+    prepareTile: ?(tileID: { z: number, x: number, y: number }) => ?T,
+    unloadTile: ?(tileID: { z: number, x: number, y: number }) => void,
+    onAdd: ?(map: Map, callback: Callback<void>) => void,
```

Comment:
I think we should also remove the callback here and just have it be `onAdd: ?(map: Map) => void`

Reply:
We need the callback here in cases when the user needs to prepare something async, e.g., resolve custom TileJSON endpoint. After the user called the callback, the CustomSource will be in the `loaded` state

Reply:
I think we should assume things like that are handled by the user before adding the source. Currently it wouldn't be possible to update any of the source's properties using that tilejson since all the properties are copied in the constructor.

If we do need something like this, maybe it should be a separate method. onAdd doesn't take a callback in either `CustomLayerInterface` or `IControl` (ui control interface)